File: gemfire-core/src/main/java/com/gemstone/gemfire/internal/cache/tx/TransactionalOperation.java
/*=========================================================================
 * Copyright (c) 2010-2014 Pivotal Software, Inc. All Rights Reserved.
 * This product is protected by U.S. and international copyright
 * and intellectual property laws. Pivotal products are covered by
 * one or more patents listed at http://www.pivotal.io/patents.
 *=========================================================================
 */
/**
 * 
 */
package com.gemstone.gemfire.internal.cache.tx;

import java.util.Collection;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

/**
 * @author Bruce Schuchardt
 *
 */
public class TransactionalOperation {
  static public enum ServerRegionOperation {
    CONTAINS_KEY,
    CONTAINS_VALUE,
    CONTAINS_VALUE_FOR_KEY,
    DESTROY,  // includes REMOVE(k,v)
    EXECUTE_FUNCTION,
    GET,
    GET_ENTRY,
    GET_ALL,
    INVALIDATE,
    KEY_SET,
    PUT, // includes PUT_IF_ABSENT
    PUT_ALL,
    REMOVE_ALL;
    
    /**
     * @param op
     * @return true if the key associated with this op should be locked on
     * clients locally; false otherwise
     */
    public static boolean lockKeyForTx(ServerRegionOperation op) {
      if (op == PUT || op == PUT_ALL || op == REMOVE_ALL || op == DESTROY || op == INVALIDATE) {
        return true;
      }
      return false;
    }
    };

//  private ClientTXStateStub clienttx;
  private ServerRegionOperation operation;
  private String regionName;
  private Object key;
  private Object[] arguments;

  protected TransactionalOperation(ClientTXStateStub clienttx, String regionName, ServerRegionOperation op, Object key, Object arguments[]) {
//    this.clienttx = clienttx;
    this.regionName = regionName;
    this.operation = op;
    this.key = key;
    this.arguments = arguments;
  }
  
  /**
   * @return the name of the affected region
   */
  public String getRegionName() {
    return this.regionName;
  }
  
  public ServerRegionOperation getOperation() {
    return this.operation;
  }
  
  public Object getKey() {
    return this.key;
  }
  
  /**
   * @return set of keys if the operation was PUTALL or REMOVEALL; null otherwise
   */
  public Set<Object> getKeys() {
    if (this.operation == ServerRegionOperation.PUT_ALL) {
      Map m = (Map) this.arguments[0];
      return m.keySet();
    } else if (this.operation == ServerRegionOperation.REMOVE_ALL) {
      Collection<Object> keys = (Collection<Object>) this.arguments[0];
      return new HashSet(keys);
    }
    return null;
  }
  
  /** this returns internal arguments for reinvoking methods on the ServerRegionProxy */
  public Object[] getArguments() {
    return this.arguments;
  }
  
  @Override
  public String toString() {
    return "TransactionalOperation(region="+this.regionName+"; op="+this.operation
     + "; key=" + this.key + ")";
  }

}

File: src/cppcache/src/TransactionalOperation.cpp
/*=========================================================================
 * Copyright (c) 2010-2014 Pivotal Software, Inc. All Rights Reserved.
 * This product is protected by U.S. and international copyright
 * and intellectual property laws. Pivotal products are covered by
 * one or more patents listed at http://www.pivotal.io/patents.
 *=========================================================================
 */
/*
 * TransactionalOperation.cpp
 *
 *  Created on: 10-May-2011
 *      Author: ankurs
 */

#include "TransactionalOperation.hpp"
#include <gfcpp/CacheableString.hpp>
#include <gfcpp/Cache.hpp>
#include <gfcpp/FunctionService.hpp>

namespace gemfire {

TransactionalOperation::TransactionalOperation(ServerRegionOperation op,
                                               const char* regionName,
                                               CacheableKeyPtr key,
                                               VectorOfCacheablePtr arguments)
    : m_operation(op),
      m_regionName(regionName),
      m_key(key),
      m_arguments(arguments) {}

TransactionalOperation::~TransactionalOperation() {}

CacheablePtr TransactionalOperation::replay(Cache* cache) {
  CacheablePtr result = NULLPTR;

  switch (m_operation) {
    case GF_CONTAINS_KEY:
      result = CacheableBoolean::create(
          cache->getRegion(m_regionName)->containsKeyOnServer(m_key));
      break;
    case GF_CONTAINS_VALUE:
      GfErrTypeThrowException("Contains value not supported in transaction",
                              GF_NOTSUP);
      // result = Boolean.valueOf(containsValue(args[0], true));
      break;
    case GF_CONTAINS_VALUE_FOR_KEY:
      result = CacheableBoolean::create(
          cache->getRegion(m_regionName)->containsValueForKey(m_key));
      break;
    case GF_DESTROY:  // includes REMOVE(k,v)
      cache->getRegion(m_regionName)->destroy(m_key, m_arguments->at(0));
      break;
    case GF_EXECUTE_FUNCTION: {
      ExecutionPtr execution;
      if (m_regionName == NULL) {
        execution = FunctionService::onServer(CachePtr(cache));
      } else {
        execution = FunctionService::onRegion(cache->getRegion(m_regionName));
      }
      result =
          execution->withArgs(m_arguments->at(0))
              ->withFilter(m_arguments->at(1))
              ->withCollector(m_arguments->at(2))
              ->execute(m_arguments->at(3)->toString()->asChar(),
                        (static_cast<CacheableInt32Ptr>(m_arguments->at(4)))
                            ->value());
    } break;
    case GF_GET:
      result = cache->getRegion(m_regionName)->get(m_key, m_arguments->at(0));
      break;
    case GF_GET_ENTRY:
      GfErrTypeThrowException("getEntry not supported in transaction",
                              GF_NOTSUP);
      break;
    case GF_GET_ALL:
      cache->getRegion(m_regionName)
          ->getAll(
              *(static_cast<VectorOfCacheableKeyPtr>(m_arguments->at(0))).ptr(),
              m_arguments->at(1), m_arguments->at(2),
              (static_cast<CacheableBooleanPtr>(m_arguments->at(3)))->value());
      break;
    case GF_INVALIDATE:
      cache->getRegion(m_regionName)->invalidate(m_key, m_arguments->at(0));
      break;
    case GF_REMOVE:
      cache->getRegion(m_regionName)
          ->remove(m_key, m_arguments->at(0), m_arguments->at(1));
      break;
    case GF_KEY_SET:
      cache->getRegion(m_regionName)
          ->serverKeys(
              *(static_cast<VectorOfCacheableKeyPtr>(m_arguments->at(0)))
                   .ptr());
      break;
    case GF_CREATE:  // includes PUT_IF_ABSENT
      cache->getRegion(m_regionName)
          ->create(m_key, m_arguments->at(0), m_arguments->at(1));
      break;
    case GF_PUT:  // includes PUT_IF_ABSENT
      cache->getRegion(m_regionName)
          ->put(m_key, m_arguments->at(0), m_arguments->at(1));
      break;
    case GF_PUT_ALL:
      cache->getRegion(m_regionName)
          ->putAll(
              *(static_cast<HashMapOfCacheablePtr>(m_arguments->at(0))).ptr(),
              (static_cast<CacheableInt32Ptr>(m_arguments->at(1)))->value());
      break;
    default:
      throw UnsupportedOperationException(
          "unknown operation encountered during transaction replay");
  }

  return result;
}
}  // namespace gemfire
